spec/formatting.md

```diff
@@ -222,6 +222,61 @@ the following steps are taken:
 
    The form that resolved _operand_ and _option_ values take is implementation-defined.
 
+   A _local-declaration_ binds the resolved value of an _expression_
+   to a _variable_.
+   Thus, the output of one _function_ is potentially the _operand_
+   of another _function_. In other words, formatting functions
+   compose with each other.
+   For example, in
+   ```
+   .input {$n :number minIntegerDigits=3}
+   .local {$n1 :number maxFractionDigits=3}
+   ```
+   the second call to `:number` composes with the first call.
+
+   In addition, selector functions compose with formatting functions
+   in the sense that a selector function's _operand_
+   may be the output of any formatting function.
+
+   Implementations SHOULD provide a means for formatting functions
+   to compose with each other
+   and for formatting functions to compose with selector functions.
+   Implementations that provide a means for defining custom functions
+   SHOULD provide a means for those functions to return values
+   that contain enough information
+   (e.g. the resolved _operand_ and _option_ values
+   that the function was called with)
+   to be used as inputs to subsequent function calls.
+   For example, an implementation in a typed programming language
+   MAY define an interface that custom functions implement.
+   Such an interface SHOULD define an implementation-specific
+   argument type `T` and return type `U` for custom formatting functions
+   such that `U` can be coerced to `T` without loss of information.
+   The type `U`
+   (or a type that `U` can be coerced to without loss of information)
+   SHOULD also be the input type of custom selector functions.
+
+> [!NOTE]
+> In the Tech Preview, the spec leaves the behavior of the previous
+> example implementation-dependent. Supposing that
```

Comment:
Isn't this PR still in scope of the Tech Preview? Or do you mean that it's implementation-dependent because of the `SHOULD`?

Reply:
This PR's text is part of the Tech Preview.

Reply:
We need to make clear that the 'return type' is not the 'thing bound to a variable'. The 'thing bound to a variable' also includes options.

Reply:
That's one of the questions, though. If the contract with function implementations is that they are responsible for returning a thing containing all the options they want to preserve, then whatever a function returns _is_ the thing bound to a variable (modulo possible lazy evaluation). If function implementations have no such responsibility, then yes, the formatter has to do additional processing to transform the "thing returned by a function" into the "thing bound to a variable".